### PR TITLE
Add error tcActivePatternArgumentCountNotMatch

### PR DESCRIPTION
**Pull request by Tangent-90**
## Description

Add a new error to improve readability for active pattern case argument count mismatch.

`This active pattern case needs %d argument(s) and %d return value, but here has %d argument(s) and %d return value.`

QUESTION: what is the best name for the "return value" of an active pattern case?

before:

图片

after:

图片

Fixes # (issue, if applicable)

## Checklist

- [ ] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [ ] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/releae-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

**Comment by Tangent-90** [comment]
Should we allow this? 
```fsharp
let (|T2|) a b = ()
match 2 with T2 1 -> ();;
```

图片

**Comment by vzarytovskii** [comment]
> QUESTION: what is the best name for the "return value" of an active pattern case?

I think "return value" works, however "1 return value" in the error message looks a bit awkward.

**Comment by T-Gro** [comment]
My feedback for the wording:
For parametrized active patterns, the message should say "parameters" when there is a mismatch.
For non-parametrized actived patterns (will be a majority), the 0 parameters/arguments would be better not mentioned at all in the message.

For the return value, I would use the wording:
`This active pattern returns 1 value, but the usage here matches 0 values`

**Comment by Tangent-90** [comment]
What about this?

For parametrized single case active pattern and partial active pattern：

This active pattern requires %d parameter(s) and returns %d value, but the usage here matches %d argument(s) and %d return value.

For else：

This active pattern returns %d value, but the usage here matches %d return value(s).

@vzarytovskii @T-Gro

**Comment by T-Gro** [comment]
> value(s).

That looks better to me.
But anywhere you had "%d value", I would suggest to apply "value(s)".
Otherwise it would just look bad in the case of 0,2 or more.

**Review by T-Gro — APPROVED** [review]

**Review by abonie — APPROVED** [review]

**Comment by psfinaki** [comment]
It's a good idea to improve this error. At the same time, as a non-user of active patterns, I got lost even with the new message. 
"Return value(s)" sound very misleading to me, how about "captured value(s)"?

File: src/Compiler/FSComp.txt [hunk 1]
```diff
@@ -1745,4 +1745,6 @@ featureReuseSameFieldsInStructUnions,"Share underlying fields in a [<Struct>] di
 3862,parsStaticMemberImcompleteSyntax,"Incomplete declaration of a static construct. Use 'static let','static do','static member' or 'static val' for declaration."
 3863,parsExpectingField,"Expecting record field"
 3864,tooManyMethodsInDotNetTypeWritingAssembly,"The type '%s' has too many methods. Found: '%d', maximum: '%d'"
-3865,parsOnlySimplePatternsAreAllowedInConstructors,"Only simple patterns are allowed in primary constructors"
\ No newline at end of file
+3865,parsOnlySimplePatternsAreAllowedInConstructors,"Only simple patterns are allowed in primary constructors"
+3866,tcParameterizedActivePatternArgumentCountNotMatch,"This active pattern requires %d parameter(s) and returns %d value(s), but the usage here matches %d argument(s) and %d return value(s)."
+3867,tcNoParameterActivePatternArgumentCountNotMatch,"This active pattern returns %d value(s), but the usage here matches %d return value(s)."
```
Comment (brianrourkeboll) [inline src/Compiler/FSComp.txt:1750]: These error numbers were used in #17055. Be sure to use new numbers when resolving the merge conflict.

**Comment by brianrourkeboll** [comment]
@psfinaki

> It's a good idea to improve this error. … I got lost even with the new message.

I agree.

@Tangent-90

> QUESTION: what is the best name for the "return value" of an active pattern case?

The F# spec (§7.2.3) uses language like this:

> The function accepts $n+1$ arguments, where the last argument ($inp$) is
the value to match, and … [can/must return] [any type/a value of type …]

But that's from the perspective of an active pattern's _definition_, not its use in pattern-matching.

One alternative might be to phrase the new error messages from the perspective of the _consumer_ of the active pattern — which is what I think we're trying to cater to here.

That is, you could also describe an active pattern[^1] as taking or expecting $[0,n]$ _expression arguments_ and $[0,1]$ _pattern arguments_[^2], since those are what the consumer must supply at the match site:

```fsharp
let (|P|…|) expr = P
// 0 expression arguments, 0 pattern arguments.
match expr with P -> …
```

```fsharp
let (|P|…|) expr1 = P expr2
// 0 expression arguments, 1 pattern argument.
match expr1 with P pat -> …
//                 ↑  
//                 This must be a pattern, not an expression.
```

```fsharp
let (|P|…|) expr2 expr1 = P
// 1 expression argument, 0 pattern arguments.
match expr1 with P expr2 -> …
//                 ↑   
//                 This must be an expression, not a pattern.
```

```fsharp
let (|P|…|) expr2 expr1 = P expr3
// 1 expression argument, 1 pattern argument.
match expr1 with P expr2 pat -> …
//                 ↑     ↑  
//                 These must be:
//                 (1) an expression, not a pattern; and
//                 (2) a pattern, not an expression,
//                 respectively.
```

What about something like these below? It would mean four distinct error messages, but I think they might be a bit easier to understand. What do you think @psfinaki @Tangent-90?

- `This active pattern does not expect any arguments, i.e., it should be used like 'P' instead of 'P x'.`[^3]
- `This active pattern expects exactly one pattern argument, e.g., 'P x'.`
- `This active pattern expects %d expression argument(s), e.g., 'P e1...'.`
- `This active pattern expects %d expression argument(s) and a pattern argument, e.g., 'P e1... pat'.`

```fsharp
// Expects: 0 expression arguments, 0 pattern arguments.
let (|P|…|) expr = P

match expr with P pat -> …
//                ↑
//                This active pattern does not expect any arguments, i.e., it should be used like 'P' instead of 'P x'.

match expr1 with P expr2 -> …
//                ↑
//                This active pattern does not expect any arguments, i.e., it should be used like 'P' instead of 'P x'.
```

```fsharp
// Expects: 0 expression arguments, 1 pattern argument.
let (|P|…|) expr1 = P expr2

match expr1 with P -> …
//                 ↑
//                 This active pattern expects exactly one pattern argument, e.g., 'P x'.

match expr1 with P expr2 pat-> …
//                 ↑
//                 This active pattern expects exactly one pattern argument, e.g., 'P x'.
```

```fsharp
// Expects: 1 expression argument, 0 pattern arguments.
let (|P|…|) expr2 expr1 = P

match expr1 with P -> …
//                 ↑
//                 This active pattern expects 1 expression argument(s), e.g., 'P e1...'.

match expr1 with P expr2 expr3 -> …
//                 ↑
//                 This active pattern expects 1 expression argument(s), e.g., 'P e1...'.
```

```fsharp
// Expects: 2 expression arguments, 0 pattern arguments.
let (|P|…|) expr2 expr3 expr1 = P

match expr1 with P -> …
//                 ↑
//                 This active pattern expects 2 expression argument(s), e.g., 'P e1...'.

match expr1 with P expr2 expr3 -> …
//                 ↑
//                 This active pattern expects 2 expression argument(s), e.g., 'P e1...'.
```

```fsharp
// Expects: 1 expression argument, 1 pattern argument.
let (|P|…|) expr2 expr1 = P expr3

match expr1 with P pat -> …
//                 ↑
//                 This active pattern expects 1 expression argument(s) and a pattern argument, e.g., 'P e1... pat'.

match expr1 with P expr2 expr3 pat -> …
//                 ↑
//                 This active pattern expects 1 expression argument(s) and a pattern argument, e.g., 'P e1... pat'.
```

```fsharp
// Expects: 2 expression arguments, 1 pattern argument.
let (|P|…|) expr2 expr3 expr1 = P expr4

match expr1 with P pat -> …
//                 ↑
//                 This active pattern expects 2 expression argument(s) and a pattern argument, e.g., 'P e1... pat'.

match expr1 with P expr2 expr3 expr4 pat -> …
//                 ↑
//                 This active pattern expects 2 expression argument(s) and a pattern argument, e.g., 'P e1... pat'.
```

[^1]: When used as such —

    ```fsharp
    let (KeyValue (key, value)) = keyValuePair
    ```

    — and not when used as a function like

    ```fsharp
    let key, value = (|KeyValue|) keyValuePair
    ```
[^2]: When the pattern is matched against, the last argument or "return value" — of which there is always either 0 or 1 — is always itself a _pattern_, whether named, as `x` in `P x`, or matched against in turn with a nested pattern, as `Q …` in `P (Q …)`.
[^3]: If we have easy access to the pattern case name being used, we could also interpolate it with `%s` instead of using `P`.

**Comment by psfinaki** [comment]
Ufff yeah I tend to agree in that this might be better handled with multiple errors. I read the docs for active patterns today (again) and it's true that the terminology and usage is a bit confusing at times.

Like, if I read just this
image

I am already confused what's an _argument_ here.

And then what, there are partitions, and banana clips, and only active patterns with exactly one result can accept arguments, right. 

This all is a bit unfortunate, so indeed I think the best we can do for F# users is to just point to the exact form of the expression that they should use in each case.

**Comment by Tangent-90** [comment]
> What about something like these below? It would mean three distinct error messages, but I think they might be a bit easier to understand. 

This one is also possible (for partial active patterns that return bool or warp a uint):

- `This active pattern expects %d expression argument(s), e.g., 'P e1...'.`

**Comment by brianrourkeboll** [comment]
> > What about something like these below? It would mean three distinct error messages, but I think they might be a bit easier to understand. 
> 
> This one is also possible (for partial active patterns that return bool or warp a uint):
> 
> - `This active pattern expects %d expression argument(s), e.g., 'P e1...'.`
> 

Ah, OK. I actually originally had that in my examples, but don't you otherwise run into https://github.com/dotnet/fsharp/pull/16846#issuecomment-1989522090 and need the `()` pattern after the expression?

**Comment by brianrourkeboll** [comment]
> That is not a partial active pattern. And before this PR, the usage also cannot compile without a `()`. This PR will not change that.

Right. Although it's arguably somewhat unfortunate that total active patterns differ from partial/Boolean-returning ones in this way.

These were the four messages I originally had in my comment, the third of which sounds like it would be used for partial/Boolean-returning parameterized patterns:

1. `This active pattern does not expect any arguments, i.e., it should be used like 'P' instead of 'P x'.`
2. `This active pattern expects exactly one pattern argument, e.g., 'P x'.`
3. `This active pattern expects %d expression argument(s), e.g., 'P e1...'.`
4. `This active pattern expects %d expression argument(s) and a pattern argument, e.g., 'P e1... pat'.`

I.e., (3) would be used for:

```fsharp
// Expects: 1 expression argument, 0 pattern arguments.
let (|P|_|) expr2 expr1 = Some P

match expr1 with P -> …
//               ↑
//               This active pattern expects 1 expression argument(s), e.g., 'P e1...'.

match expr1 with P expr2 expr3 -> …
//               ↑
//               This active pattern expects 1 expression argument(s), e.g., 'P e1...'.
```

**Comment by Tangent-90** [comment]
图片
the four situation @brianrourkeboll

**Comment by brianrourkeboll** [comment]
@Tangent-90

> the four situation

I think this is a pretty solid improvement over the status quo.

I guess since we _do_ have the expected number of expression and pattern arguments available to us, we could in theory even offer something like `P e1 e2 e3`, `P e1 e2 pat`, etc., in the examples, rather than using `e1...` as a stand-in for $n$ expressions.

I wonder whether it would be worth doing something like:

```fsharp
// Use a prebuilt expr args string for these that has an example with the actual number of args.
"This active pattern expects %d expression argument(s), e.g., '%s%s'."
"This active pattern expects %d expression argument(s) and a pattern argument, e.g., '%s%s pat'."
```

and

```fsharp
let fmtExprArgs paramCount =
    let rec loop i (sb: StringBuilder) =
        let cutoff = 10
        if i > paramCount then sb.ToString()
        elif i > cutoff then sb.Append("...").ToString()
        else loop (i + 1) (sb.Append(" e").Append i)

    loop 1 (StringBuilder())

let msg =
    match paramCount, returnCount with
    | 0, 0 -> FSComp.SR.tcActivePatternArgsCountNotMatchNoArgsNoPat(caseName, caseName)
    | 0, _ -> FSComp.SR.tcActivePatternArgsCountNotMatchOnlyPat(caseName)
    | _, 0 -> FSComp.SR.tcActivePatternArgsCountNotMatchArgs(paramCount, fmtExprArgs paramCount, caseName)
    | _, _ -> FSComp.SR.tcActivePatternArgsCountNotMatchArgsAndPat(paramCount, fmtExprArgs paramCount, caseName)
```

giving

```
This active pattern expects 2 expression argument(s), e.g., 'P e1 e2'.
This active pattern expects 3 expression argument(s) and a pattern argument, e.g., 'P e1 e2 e3 pat'.
```

etc.

Or would that be overkill? What do you think @psfinaki?

**Comment by psfinaki** [comment]
In the topic of diagnostics, it's hard to do an overkill, it's more about perfectionism here now :)

The current state of the art is definitely an improvement compared to the main. "Use 'IsEven' instead of 'IsEven x' probably captures most of the usecases and make the error very clear. The other errors importantly contain the expected number of "things" (arguments, expressions, patterns) and also list all these terms, like hey, there are all these capabilities of active patterns, read the docs if you're interested.

So @Tangent-90 if you're not tired yet, feel free to use Brian's addition. If you are tired, feel free to leave things as they are. I'd just ask to add the tests for the new errors, among other things they can serve as documentation for the time being.

**Comment by Tangent-90** [comment]
message updated @brianrourkeboll @psfinaki

File: src/Compiler/Checking/CheckExpressions.fs [hunk 2]
```diff
@@ -5119,22 +5119,75 @@ and TcPatLongIdentActivePatternCase warnOnUpper (cenv: cenv) (env: TcEnv) vFlags
...
-
-            if dtys.Length = args.Length + 1 &&
-                ((isOptionTy g retTy && isUnitTy g (destOptionTy g retTy)) ||
-                (isValueOptionTy g retTy && isUnitTy g (destValueOptionTy g retTy))) ||
-                // `bool` partial AP always be treated as `unit option`
-                // For `val (|P|_|) : _ -> bool`, only allow `match x with | P -> ...`
-                // For `val (|P|_|) : _ -> _ -> bool`, only allow `match x with | P parameter -> ...`
-                (not apinfo.IsTotal && isBoolTy g retTy) then
+        let rec IsNotSolved ty =
+            match ty with
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckExpressions.fs:5123]: Shall `ty` be stripped here? @T-Gro

**Review by psfinaki — APPROVED** [review]
Good stuff, thanks for this. Active patterns are much more usable now :)

File: src/Compiler/Checking/CheckExpressions.fs (hunk 2 above)
Comment (T-Gro) [reply src/Compiler/Checking/CheckExpressions.fs:5123]: Stripping would remove solved TType_var.
I think for the case of resolving solved/unsolved typars, this is good.

What is not covered and stripping would do:
- Aliases - I cannot come up with a way of placing an alias here
- UoMs - not relevant for IsNotSolved
- ByRefs - not relevant for IsNotSolved

**Review by T-Gro — APPROVED** [review]

File: src/Compiler/Checking/CheckExpressions.fs [hunk 3]
```diff
@@ -5119,22 +5119,75 @@ and TcPatLongIdentActivePatternCase warnOnUpper (cenv: cenv) (env: TcEnv) vFlags
...
+                loop 1 (Text.StringBuilder())
+
+            let caseName = apinfo.ActiveTags[idx]
+            let msg =
+                match paramCount, returnCount with
+                | 0, 0 -> FSComp.SR.tcActivePatternArgsCountNotMatchNoArgsNoPat(caseName, caseName)
+                | 0, _ -> FSComp.SR.tcActivePatternArgsCountNotMatchOnlyPat(caseName)
+                | _, 0 -> FSComp.SR.tcActivePatternArgsCountNotMatchArgs(paramCount, caseName, fmtExprArgs paramCount)
+                | _, _ -> FSComp.SR.tcActivePatternArgsCountNotMatchArgsAndPat(paramCount, caseName, fmtExprArgs paramCount)
+            error(Error(msg, m))
```
Comment (edgarfgp) [inline src/Compiler/Checking/CheckExpressions.fs:5153]: If possible we should use `errorR` to allow the rest of the analysis to continue.

Comment (Tangent-90) [reply src/Compiler/Checking/CheckExpressions.fs]: updated

Comment (Tangent-90) [reply src/Compiler/Checking/CheckExpressions.fs]: I think it should not use `errorR` here, as if this check not pass, then the compiler won't know how to apply the args to the active pattern function

**Review by edgarfgp — APPROVED** [review]

**Comment by psfinaki** [comment]
Build errors are unrelated, we hope to resolve those soon, sorry.

**Pull request merged**